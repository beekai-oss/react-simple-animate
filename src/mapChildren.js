// @flow

export default function mapChildren(
  childrenStoreInState: Object | Array<any>,
  children: Object | Array<any>,
) {
  let previousIndex: number = 0;
  const result: Array<any> = [];
  const childrenShallowCopy = Array.isArray(children) ? children : [children];
  const childrenStoreInStateShallowCopy = Array.isArray(childrenStoreInState)
    ? childrenStoreInState
    : [childrenStoreInState];

  // any children got removed
  childrenStoreInStateShallowCopy.forEach((element: Object) => {
    if (!element.hasOwnProperty('key')) return;

    if (childrenShallowCopy.find(child => child.key === element.key)) {
      result.push(element);
    } else {
      result.push({ ...element, willUnmount: true });
    }
  });

  // append missing child
  childrenShallowCopy.forEach((child, index) => {
    if (child.hasOwnProperty('key') && !result.find(r => r.key === child.key)) {
      result.splice(index === 0 ? 0 : index + 1, 0, {
        ...child,
        willMount: true,
      });
    }

    previousIndex = index;
  });

  return result;
}
