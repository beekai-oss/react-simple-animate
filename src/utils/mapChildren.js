// @flow
import React from 'react';
import propsGenerator from './propsGenerator';
import type { Props, State } from '../animate';

const filterUnMountChildren = (children: Array<React$Element<any>>) =>
  children.filter((child: Object) => !child.willUnmount);

export default function mapChildren(props: Props, state: State) {
  const { childrenStoreInState, willLeave, willEnter } = state;

  if (!Array.isArray(childrenStoreInState) || !Array.isArray(props.children)) {
    return null;
  }

  let children = childrenStoreInState;

  if (willLeave) {
    children = filterUnMountChildren(childrenStoreInState);
  } else if (willEnter) {
    children = props.children;
  }

  return children.map((child: Object) => {
    if (!child) return null;

    let componentProps = {};
    let startAnimation = true;

    const { willMount = false, willUnmount = false } = child;
    const isAnimateComponent =
      child.type &&
      child.type.displayName &&
      child.type.displayName === 'ReactSimpleAnimate';

    if (isAnimateComponent) {
      if (willMount) {
        startAnimation = willEnter;
      } else if (willUnmount) {
        startAnimation = willLeave;
      }
    } else {
      componentProps = propsGenerator(
        {
          ...props,
        },
        state,
        {
          willUnmount,
          willMount,
        },
      );
    }

    return React.cloneElement(child, {
      ...{ ...child.props },
      ...(!isAnimateComponent && componentProps.style
        ? { style: componentProps.style }
        : { startAnimation }),
    });
  });
}
