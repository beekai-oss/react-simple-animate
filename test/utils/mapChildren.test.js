import mapChildren from '../../src/utils/mapChildren';
import Animate from '../../src/animate';
import React from 'react';

describe('mapChildren', () => {
  it('should return null when childrenStoreInState is null', () => {
    expect(
      mapChildren(
        {
          children: null,
        },
        {
          childrenStoreInState: null,
        },
      ),
    ).toBe(null);
  });

  describe('when willLeave set to true', () => {
    it('should filter out components which will unmont', () => {
      expect(
        mapChildren(
          {},
          {
            willLeave: true,
            childrenStoreInState: [
              {
                willUnmount: true,
              },
            ],
          },
        ),
      ).toMatchSnapshot();
    });
  });

  describe('when willLeave set to false', () => {
    it('should not filter components', () => {
      expect(
        mapChildren(
          {},
          {
            willLeave: false,
            childrenStoreInState: [
              {
                willUnmount: true,
              },
            ],
          },
        ),
      ).toMatchSnapshot();
    });
  });

  describe('when children is Animate component', () => {
    const animateComponent = <Animate />;

    it('should apply startAnimation', () => {
      expect(
        mapChildren(
          {},
          {
            childrenStoreInState: [{ ...animateComponent }],
          },
        ),
      ).toMatchSnapshot();
    });

    it('should apply startAnimation according to willEnter state', () => {
      expect(
        mapChildren(
          {},
          {
            willEnter: false,
            childrenStoreInState: [{ ...animateComponent, willMount: true }],
          },
        ),
      ).toMatchSnapshot();

      expect(
        mapChildren(
          {},
          {
            willEnter: true,
            childrenStoreInState: [{ ...animateComponent, willMount: true }],
          },
        ),
      ).toMatchSnapshot();
    });

    it('should apply startAnimation according to willLeave state', () => {
      expect(
        mapChildren(
          {},
          {
            willLeave: false,
            childrenStoreInState: [{ ...animateComponent, willUnmount: true }],
          },
        ),
      ).toMatchSnapshot();

      expect(
        mapChildren(
          {},
          {
            willLeave: true,
            childrenStoreInState: [{ ...animateComponent, willUnmount: true }],
          },
        ),
      ).toMatchSnapshot();
    });
  });
});
