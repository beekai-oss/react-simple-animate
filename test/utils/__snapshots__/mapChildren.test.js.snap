// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapChildren when children is Animate component should apply startAnimation 1`] = `
Array [
  <ReactSimpleAnimate
    startAnimation={true}
/>,
]
`;

exports[`mapChildren when children is Animate component should apply startAnimation according to willEnter state 1`] = `
Array [
  <ReactSimpleAnimate
    startAnimation={false}
/>,
]
`;

exports[`mapChildren when children is Animate component should apply startAnimation according to willEnter state 2`] = `
Array [
  <ReactSimpleAnimate
    startAnimation={true}
/>,
]
`;

exports[`mapChildren when children is Animate component should apply startAnimation according to willLeave state 1`] = `
Array [
  <ReactSimpleAnimate
    startAnimation={false}
/>,
]
`;

exports[`mapChildren when children is Animate component should apply startAnimation according to willLeave state 2`] = `Array []`;

exports[`mapChildren when willLeave set to false should not filter components 1`] = `
Array [
  <Unknown
    style={
        Object {
            "transition": "0.3s all linear",
          }
    }
/>,
]
`;

exports[`mapChildren when willLeave set to true should filter out components which will unmont 1`] = `Array []`;
